File: server/server-core/src/main/java/com/simple/castle/server/composition/BaseObject.java
package com.simple.castle.server.composition;

import com.badlogic.gdx.graphics.g3d.ModelInstance;
import com.badlogic.gdx.utils.Disposable;
import com.simple.castle.server.physic.unit.PhysicObject;

public class BaseObject implements Disposable {
    private final String id;
    private final PhysicObject physicObject;
    private final ModelInstance modelInstance;
    private Boolean hide;

    public BaseObject(Constructor constructor) {
        this.id = constructor.getId();

        this.physicObject = constructor.buildPhysic();
        this.modelInstance = constructor.buildModel();

        this.physicObject.getMotionState().setWorldTransform(modelInstance.transform);

        this.hide = constructor.getHide();
    }

    public String getId() {
        return id;
    }

    public ModelInstance getModelInstance() {
        return modelInstance;
    }

    public PhysicObject getPhysicObject() {
        return physicObject;
    }

    public Boolean getHide() {
        return hide;
    }

    public void setHide(Boolean hide) {
        this.hide = hide;
    }

    @Override
    public void dispose() {
        physicObject.dispose();
    }
}

File: server/server-core/src/main/java/com/simple/castle/server/kt/composition/BaseObject.kt
package com.simple.castle.server.kt.composition

import com.badlogic.gdx.graphics.g3d.ModelInstance
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.utils.Disposable
import com.simple.castle.server.kt.physic.PhysicObject
import java.util.*

@Suppress("LeakingThis")
open class BaseObject(constructor: Constructor) : Disposable {
    val nodeName: String = constructor.nodeName
    val id: String = UUID.randomUUID().toString()
    val physicObject: PhysicObject = constructor.buildPhysic()
    val modelInstance: ModelInstance = constructor.buildModel()
    var hide: Boolean

    init {
        physicObject.motionState.transform = modelInstance.transform
        physicObject.body.motionState = physicObject.motionState
        hide = constructor.hide
        if (constructor.mass != 0.0f) {
            physicObject.body.collisionShape.calculateLocalInertia(constructor.mass, Vector3.Zero)
        }
        physicObject.body.userData = this
    }

    override fun dispose() {
        physicObject.dispose()
    }
}